core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java

```diff
@@ -97,75 +133,117 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
+       */
+      if (!inclusive) {
+        for (Entry<ByteSequence,LocalityGroup> entry : groups.groupByCf.entrySet()) {
+          if (!cfSet.contains(entry.getKey())) {
+            groupsToUse.add(entry.getValue());
+          }
+        }
+      } else if (groups.groupByCf.size() <= cfSet.size()) {
+        for (Entry<ByteSequence,LocalityGroup> entry : groups.groupByCf.entrySet()) {
+          if (cfSet.contains(entry.getKey())) {
+            groupsToUse.add(entry.getValue());
+          }
         }
       } else {
-        /*
-         * Need to consider the following cases for inclusive and exclusive (lgcf:locality group column family set, cf:column family set) lgcf and cf are
-         * disjoint lgcf and cf are the same cf contains lgcf lgcf contains cf lgccf and cf intersect but neither is a subset of the other
-         */
-
-        for (Entry<ByteSequence,MutableLong> entry : lgr.columnFamilies.entrySet())
-          if (entry.getValue().longValue() > 0)
-            if (cfSet.contains(entry.getKey())) {
-              if (inclusive)
-                include = true;
-            } else if (!inclusive) {
-              include = true;
-            }
+        for (ByteSequence cf : cfSet) {
+          LocalityGroup group = groups.groupByCf.get(cf);
+          if (group != null) {
+            groupsToUse.add(group);
+          }
+        }
       }
+    }
 
-      if (include) {
-        lgr.getIterator().seek(range, EMPTY_CF_SET, false);
-        hiter.addSource(lgr.getIterator());
-        numLGSeeked++;
-      }// every column family is excluded, zero count, or not present
+    for (LocalityGroup lgr : groupsToUse) {
+      lgr.getIterator().seek(range, EMPTY_CF_SET, false);
+      hiter.addSource(lgr.getIterator());
+      numLGSeeked++;
+    }
+
+    if (used != null) {
+      used.addAll(groupsToUse);
     }
 
     return numLGSeeked;
   }
 
   @Override
   public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
-    seek(this, groups, nonDefaultColumnFamilies, range, columnFamilies, inclusive);
+    Set<ByteSequence> cfSet;
+    if (columnFamilies.size() > 0)
+      if (columnFamilies instanceof Set<?>) {
+        cfSet = (Set<ByteSequence>) columnFamilies;
+      } else {
+        cfSet = new HashSet<>();
+        cfSet.addAll(columnFamilies);
+      }
+    else
+      cfSet = Collections.emptySet();
+
+    if (lastUsed != null && cfSet.equals(lastColumnFamilies) && inclusive == lastInclusive) {
+      clear();
+      for (LocalityGroup lgr : lastUsed) {
+        lgr.getIterator().seek(range, EMPTY_CF_SET, false);
+        addSource(lgr.getIterator());
+      }
+    } else {
+      if (lastUsed == null) {
+        lastUsed = new ArrayList<LocalityGroup>(lgContext.groups.length);
+      } else {
+        lastUsed.clear();
+      }
+      lastColumnFamilies = cfSet;
```

Comment:
Need to be careful with this, it could be a ref to a set a user passed.  The user could call `seek(setA)` then modify `setA` and then call `seek(setA)`.  In this use case it seems like the wrong loc groups could be used.